lib/hocon/cli.rb

```diff
@@ -0,0 +1,186 @@
+require 'optparse'
+require 'hocon'
+require 'hocon/config_render_options'
+require 'hocon/config_factory'
+require 'hocon/config_value_factory'
+require 'hocon/parser/config_document_factory'
+require 'hocon/config_error'
+
+module Hocon::CLI
+  # Aliases
+  ConfigMissingError = Hocon::ConfigError::ConfigMissingError
+  ConfigWrongTypeError = Hocon::ConfigError::ConfigWrongTypeError
+
+  # List of valid subcommands
+  SUBCOMMANDS = ['get', 'set', 'unset']
+
+  # Parses the command line flags and argument
+  # Returns a options hash with values for each option and argument
+  def self.parse_args(args)
+    options = {}
+    opt_parser = OptionParser.new do |opts|
+      subcommands = SUBCOMMANDS.join(',')
+      opts.banner = "Usage: hocon [--file HOCON_FILE] {#{subcommands}} PATH [VALUE]"
+
+      in_file_description = 'HOCON file to read/modify. If omitted, STDIN assumed'
+      opts.on('-i', '--in-file HOCON_FILE', in_file_description) do |in_file|
+        options[:in_file] = in_file
+      end
+
+      out_file_description = 'File to be written to. If omitted, STDOUT assumed'
+      opts.on('-o', '--out-file HOCON_FILE', out_file_description) do |out_file|
+        options[:out_file] = out_file
+      end
+
+      json_description = "Output values from the 'get' subcommand in json format"
+      opts.on('-j', '--json', json_description) do |json|
+        options[:json] = json
+      end
+
+      opts.on_tail('-h', '--help', 'Show this message') do
+        puts opts
+        exit
+      end
+
+      opts.on_tail('-v', '--version', 'Show version') do
+        puts Gem.loaded_specs['hocon'].version
+        exit
+      end
+    end
+    # parse! returns the argument list minus all the flags it found
+    remaining_args = opt_parser.parse!(args)
+
+    no_subcommand_error(opt_parser) unless remaining_args.size > 0
+
+    # Assume the first arg is the subcommand
+    subcommand = remaining_args.shift
+    options[:subcommand] = subcommand
+
+    case subcommand
+      when 'set'
+        subcommand_arguments_error(subcommand, opt_parser) unless remaining_args.size >= 2
+        options[:path] = remaining_args.shift
+        options[:new_value] = remaining_args.shift
+
+      when 'get', 'unset'
+        subcommand_arguments_error(subcommand, opt_parser) unless remaining_args.size >= 1
+        options[:path] = remaining_args.shift
+
+      else
+        invalid_subcommand_error(subcommand, opt_parser)
+    end
+
+    options
+  end
+
+  # Main entry point into the script
+  # Calls the appropriate subcommand
+  def self.main(opts)
+    hocon_text = get_hocon_file(opts[:in_file])
```

Comment:
This is a pretty cool tool. Any chance we could get it to flock the file to reduce race conditions if multiple processes try to manipulate it?